### PR TITLE
jit: support for various float operations part2

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: wasm/jit/jit_amd64.go [hunk 1]
```diff
@@ -1643,6 +1865,267 @@ func (c *amd64Compiler) compileSqrt(o *wazeroir.OperationSqrt) error {
...
+		err = c.emitSimpleIntToFloatConversion(x86.ACVTSQ2SD) // = CVTSI2SD for 64bit int.
+	} else if o.OutputType == wazeroir.Float32 && o.InputType == wazeroir.SignedUint64 {
+		err = c.emitUnsignedInt64ToFloatConversion(true)
+	} else if o.OutputType == wazeroir.Float64 && o.InputType == wazeroir.SignedUint64 {
+		err = c.emitUnsignedInt64ToFloatConversion(false)
+	}
+	return
+}
+
+func (c *amd64Compiler) emitUnsignedInt64ToFloatConversion(isFloat32bit bool) error {
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:1918]: without compiler explorer I couldn't have been able to finish this :D

**Comment by mathetake** [comment]
Ok the only thing left is truncation from float to int!

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -1643,6 +1865,267 @@ func (c *amd64Compiler) compileSqrt(o *wazeroir.OperationSqrt) error {
...
+	mov.From.Type = obj.TYPE_REG
+	mov.From.Reg = target.register
+	mov.To.Type = obj.TYPE_REG
+	mov.To.Reg = target.register
+	c.addInstruction(mov)
+	return nil
+}
+
+func (c *amd64Compiler) compileITruncFromF(o *wazeroir.OperationITruncFromF) error {
+	return nil
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:1885]: I will complete ITruncFromF in another PR as it should be fairly complex and would double the size of this PR :D

File: wasm/jit/jit_amd64_test.go [hunk 3]
```diff
@@ -3877,6 +3887,192 @@ func TestAmd64Compiler_compileRem(t *testing.T) {
...
+			// Generate and run the code under test.
+			code, _, err := compiler.generate()
+			require.NoError(t, err)
+			eng := newEngine()
+			jitcall(uintptr(unsafe.Pointer(&code[0])), uintptr(unsafe.Pointer(eng)), 0)
+
+			// Check the result.
+			require.Equal(t, uint64(1), eng.stackPointer)
+			if math.IsNaN(v) {
+				require.True(t, math.IsNaN(float64(math.Float32frombits(uint32(eng.stack[eng.stackPointer-1])))))
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:3927]: haha quite a mouthful.  if any part of this conversion is common, maybe extract a function.. perhaps the `math.Float32frombits(uint32` as `float32frombits`

File: wasm/jit/jit_amd64_test.go [hunk 4]
```diff
@@ -3877,6 +3887,192 @@ func TestAmd64Compiler_compileRem(t *testing.T) {
...
+						require.Equal(t, exp, actual)
+					} else if tc.outputType == wazeroir.Float32 && tc.inputType == wazeroir.SignedUint64 {
+						exp := float32(v)
+						actual := math.Float32frombits(uint32(actualBits))
+						require.Equal(t, exp, actual)
+					} else if tc.outputType == wazeroir.Float64 && tc.inputType == wazeroir.SignedUint64 {
+						exp := float64(v)
+						actual := math.Float64frombits(actualBits)
+						require.Equal(t, exp, actual)
+					}
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:4069]: else t.Fail probably

**Review by codefromthecrypt — APPROVED** [review]
Wow looks like this was both a lot of work and a lot of reading to figure out the work to do. Thanks for putting in that brain and effort!

File: wasm/jit/jit_amd64_test.go (hunk 3 above)
Comment (mathetake) [reply wasm/jit/jit_amd64_test.go]: done https://github.com/tetratelabs/wazero/pull/121/commits/f495b1d81c633831c6078cbd32a03a3122cbc3cc

**Comment by mathetake** [comment]
Thanks for the look!!!!!!!

**Pull request merged**